From be9814155ba119521709a506de223099a76a52ed Mon Sep 17 00:00:00 2001
From: Hai Pham <hai.pham.ud@renesas.com>
Date: Sat, 10 May 2025 17:24:35 +0700
Subject: [PATCH 08/46] pinctrl: renesas: Allow drive strength configuration
 via SoC driver

Allow drive strength configuration via SoC driver, which is needed for
R-Car Gen5 family drive strength configuration.

Signed-off-by: Hai Pham <hai.pham.ud@renesas.com>
---
 drivers/pinctrl/renesas/pinctrl.c | 6 +++++-
 drivers/pinctrl/renesas/sh_pfc.h  | 2 ++
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/pinctrl/renesas/pinctrl.c b/drivers/pinctrl/renesas/pinctrl.c
index f84c61eab154..210452276029 100644
--- a/drivers/pinctrl/renesas/pinctrl.c
+++ b/drivers/pinctrl/renesas/pinctrl.c
@@ -693,7 +693,11 @@ static int sh_pfc_pinconf_set(struct pinctrl_dev *pctldev, unsigned _pin,
 				pinconf_to_config_argument(configs[i]);
 			int ret;
 
-			ret = sh_pfc_pinconf_set_drive_strength(pfc, _pin, arg);
+			if (!pfc->info->ops || !pfc->info->ops->set_drive_strength)
+				ret = sh_pfc_pinconf_set_drive_strength(pfc, _pin, arg);
+			else
+				ret = pfc->info->ops->set_drive_strength(pfc, _pin, arg);
+
 			if (ret < 0)
 				return ret;
 
diff --git a/drivers/pinctrl/renesas/sh_pfc.h b/drivers/pinctrl/renesas/sh_pfc.h
index 0a7343e0a83a..72d2a82523c5 100644
--- a/drivers/pinctrl/renesas/sh_pfc.h
+++ b/drivers/pinctrl/renesas/sh_pfc.h
@@ -275,6 +275,8 @@ struct sh_pfc_soc_operations {
 			 unsigned int bias);
 	int (*pin_to_pocctrl)(unsigned int pin, u32 *pocctrl);
 	int (*pin_to_portcr)(unsigned int pin);
+	int (*set_drive_strength)(struct sh_pfc *pfc, unsigned int pin,
+				  u16 strength);
 };
 
 struct sh_pfc_soc_info {
-- 
2.34.1

